File: Graphs/ShortestReachInAGraph.cpp
#include <bits/stdc++.h>
using namespace std;

class Graph {
    public:
        unordered_map<int,vector<int> > m;
        vector<int> dist;
        vector<bool> visit;

        Graph(int n) {
            dist.resize(n,-1);
            visit.resize(n,0);

            for(int i=0;i<n;i++){
                m.insert(make_pair(i,vector<int>()));
            }
        }

        void add_edge(int u, int v) {
            m[u].push_back(v);
            m[v].push_back(u);
        }

        vector<int> shortest_reach(int start) {
            queue< pair<int,int> > q;
            q.push(make_pair(start,0));
            int e,d,i;
            visit[start]=1;

            while(!q.empty()){
                e=q.front().first;
                d=q.front().second;

                q.pop();

                for(i=0;i<m[e].size();i++){
                    if(visit[(m[e])[i]]==0){
                        visit[(m[e])[i]]=1;
                        dist[(m[e])[i]] = (d+1)*6;
                        q.push(make_pair( (m[e])[i],d+1 ));
                    }
                }
            }
            return dist;
        }

};

int main() {
    int queries;
    cin >> queries;

    for (int t = 0; t < queries; t++) {

      int n, m;
        cin >> n;
        // Create a graph of size n where each edge weight is 6:
        Graph graph(n);
        cin >> m;
        // read and set edges
        for (int i = 0; i < m; i++) {
            int u, v;
            cin >> u >> v;
            u--, v--;
            // add each edge to the graph
            graph.add_edge(u, v);
        }
      int startId;
        cin >> startId;
        startId--;
        // Find shortest reach from node s
        vector<int> distances = graph.shortest_reach(startId);

        for (int i = 0; i < distances.size(); i++) {
            if (i != startId) {
                cout << distances[i] << " ";
            }
        }
        cout << endl;
    }

    return 0;
}

File: Graphs/Shortest_Reach_in_a_Graph.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
    public static class Graph {

        private List<List<Integer>> adjList;
        int V;

        public Graph(int size) {
            this.V = size;
            adjList = new ArrayList();
            for(int i=0; i < V; i++) adjList.add(new ArrayList());
        }

        public void addEdge(int first, int second) {
            adjList.get(first).add(second);
            adjList.get(second).add(first);
        }

        public int[] shortestReach(int startId) { // 0 indexed

            boolean[] marked = new boolean[V];
            int[] distTo = new int[V];
            Arrays.fill(distTo, -1);

            Deque<Integer> queue = new ArrayDeque();
            queue.add(startId);
            marked[startId] = true;
            distTo[startId] = 0;

            while(!queue.isEmpty()){
                int v = queue.poll();
                for(int w : adjList.get(v)){
                    if(!marked[w]){
                        marked[w] = true;
                        distTo[w] = distTo[v] + 6;
                        queue.add(w);
                    }
                }
            }

            return distTo;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int queries = scanner.nextInt();

        for (int t = 0; t < queries; t++) {

            // Create a graph of size n where each edge weight is 6:
            Graph graph = new Graph(scanner.nextInt());
            int m = scanner.nextInt();

            // read and set edges
            for (int i = 0; i < m; i++) {
                int u = scanner.nextInt() - 1;
                int v = scanner.nextInt() - 1;

                // add each edge to the graph
                graph.addEdge(u, v);
            }

            // Find shortest reach from node s
            int startId = scanner.nextInt() - 1;
            int[] distances = graph.shortestReach(startId);

            for (int i = 0; i < distances.length; i++) {
                if (i != startId) {
                    System.out.print(distances[i]);
                    System.out.print(" ");
                }
            }
            System.out.println();
        }

        scanner.close();
    }
}
